### DISCUSSION TITLE
Use custom IoC providers

### DISCUSSION THREAD

**Discussion by godrose** [dotnet/maui#1284] [Q&A]
Right now for existing .NET apps I use custom IoC containers like simple container, autofac, unity, etc.
How does this scenario work in Maui bootstrapping pipeline?

**Comment by dansiegel** [comment]
There are two steps to this. You can check out [Prism.Maui](https://github.com/dansiegel/prism.maui) for an example of how to do this. 

The first thing you need is to implement `IServiceProviderFactory<T>` which you can probably get as a support package from containers that work with AspNetCore already. The second thing you need is to call the `UseServiceProviderFactory` on the builder.  For reference here is the code that we use for Prism.Maui to accomplish this.

```cs
builder.UseMauiApp<T>()
       .UseServiceProviderFactory(new PrismServiceProviderFactory(options.Container));


public class PrismServiceProviderFactory : IServiceProviderFactory<IContainerExtension>    
{
    private IContainerExtension _container { get; }
    public PrismServiceProviderFactory(IContainerExtension container)
    {
        _container = container;
    }

    public IContainerExtension CreateBuilder(IServiceCollection services)
    {
       _container.Populate(services); 
       return _container;
    }

    public IServiceProvider CreateServiceProvider(IContainerExtension containerExtension)
    {
        return containerExtension.CreateServiceProvider();
    }    
}
```
